crates/lib/src/kernel.rs

```diff
@@ -1,51 +1,269 @@
+use std::borrow::Cow;
+
 use anyhow::Result;
-use fn_error_context::context;
 
 /// This is used by dracut.
-pub(crate) const INITRD_ARG_PREFIX: &str = "rd.";
+pub(crate) const INITRD_ARG_PREFIX: &[u8] = b"rd.";
 /// The kernel argument for configuring the rootfs flags.
-pub(crate) const ROOTFLAGS: &str = "rootflags=";
-
-/// Parse the kernel command line.  This is strictly
-/// speaking not a correct parser, as the Linux kernel
-/// supports quotes.  However, we don't yet need that here.
-///
-/// See systemd's code for one userspace parser.
-#[context("Reading /proc/cmdline")]
-pub(crate) fn parse_cmdline() -> Result<Vec<String>> {
-    let cmdline = std::fs::read_to_string("/proc/cmdline")?;
-    let r = cmdline
-        .split_ascii_whitespace()
-        .map(ToOwned::to_owned)
-        .collect();
-    Ok(r)
-}
-
-/// Return the value for the string in the vector which has the form target_key=value
-pub(crate) fn find_first_cmdline_arg<'a>(
-    args: impl Iterator<Item = &'a str>,
-    target_key: &str,
-) -> Option<&'a str> {
-    args.filter_map(|arg| {
-        if let Some((k, v)) = arg.split_once('=') {
-            if target_key == k {
-                return Some(v);
+pub(crate) const ROOTFLAGS: &[u8] = b"rootflags";
+
+const DASH: u8 = 0x2d;
+const UNDERSCORE: u8 = 0x5f;
+const DOUBLE_QUOTE: u8 = 0x22;
+const EQUALS: u8 = 0x3d;
+
+pub(crate) struct Cmdline<'a>(Cow<'a, [u8]>);
+
+impl<'a, T: AsRef<[u8]> + ?Sized> From<&'a T> for Cmdline<'a> {
+    fn from(input: &'a T) -> Self {
+        Self(Cow::Borrowed(input.as_ref()))
+    }
+}
+
+impl<'a> Cmdline<'a> {
+    pub fn from_proc() -> Result<Self> {
+        Ok(Self(Cow::Owned(std::fs::read("/proc/cmdline")?)))
+    }
+
+    pub fn iter(&'a self) -> impl Iterator<Item = Parameter<'a>> {
+        let mut in_quotes = false;
+
+        self.0
+            .split(move |c| {
+                if *c == DOUBLE_QUOTE {
+                    in_quotes = !in_quotes;
+                }
+                !in_quotes && c.is_ascii_whitespace()
+            })
+            .map(Parameter::from)
+    }
+}
+
+#[derive(Debug, Eq)]
+pub(crate) struct Parameter<'a> {
+    pub key: &'a [u8],
+    pub value: Option<&'a [u8]>,
+}
+
+impl<'a> Parameter<'a> {
+    pub fn key_lossy(&self) -> String {
+        String::from_utf8_lossy(self.key).to_string()
+    }
+
+    pub fn value_lossy(&self) -> String {
+        if let Some(v) = self.value {
+            String::from_utf8_lossy(v).to_string()
+        } else {
+            String::from("")
+        }
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> From<&'a T> for Parameter<'a> {
+    fn from(input: &'a T) -> Self {
+        let input = input.as_ref();
+        let equals = input.iter().position(|b| *b == EQUALS);
```

Comment:
I think it's a lot cleaner to use `split_once` for this. There are several other examples in the code.

Reply:
`split_once` is nightly-only for slices though :frowning_face: 

https://doc.rust-lang.org/std/primitive.slice.html#method.split_once